### DISCUSSION TITLE
Firmware

### DISCUSSION THREAD

**Discussion by Gao0130** [Schievel1/dactyl_manuform_r_track#6] [Q&A] [answered]
EN：Can the author provide the firmware of the left and right directly? The reason is that my software is a bit bad. I am not very proficient in generating firmware according to the official qmk, and I have never used arduion micro and pwm 3360 sensors. So I plan to do the same wiring as the author, and I can use the author's firmware. By the way, arduion micro is also the programming method of atmega32u4 chip, you can also use qmk tool

CN：原因就是我软件方面有点垃圾，根据qmk官方生成固件不太熟练，再加上没用过arduion micro与pwm 3360传感器。所以我打算和作者同样的接线，可以使用作者的固件。顺便问一下，arduion micro也是atmega32u4芯片烧录方式也可以使用qmk tool吧

**Comment by Schievel1** [comment, marked as answer]
Hm I see. I can compile a firmware, but if you want to customize it to your taste you will have to do it yourself. 

What do you need? US-QWERTY I guess?

> By the way, arduion micro is also the programming method of atmega32u4 chip, you can also use qmk tool

This, exactly. Since they merged my PR (https://github.com/qmk/qmk_firmware/pull/17199) there is a Arduino Micro version for the tractyl manuform keyboard in official QMK. I haven't tried it, but for a basic firmware you should be able to even use [QMK Configurator](https://config.qmk.fm/#/handwired/tractyl_manuform/5x6_right/arduinomicro/LAYOUT_5x6_right).
You won't have those extra features of the trackball I describe [here](https://github.com/Schievel1/dactyl_manuform_r_track#a-word-about-trackball-modes), though.

**Reply by Gao0130** [reply]
Thank you very much for helping me provide the firmware. For this keyboard, I only need to change the function of the via key. As for the various working modes of the trackball, I only need the default. It is just used as an imitation of the mouse function, and it is not too much to use. Frequently, it would be great if a function similar to modifying keys could be implemented.

**Reply by Schievel1** [reply]
Sorry, I don't understand. What is a 'via' key?
If you only need the basic functions, I would recommend you give that QMK Configurator a try. It is very easy to use and you would get a hex file directly out of it.

**Reply by Gao0130** [reply]
EN：Can't this keyboard change keys in real time?  www.caniusevia.com.  The above URL is via software official website.  The left and right hand firmware with the .hex suffix is ​​used for burning, and the via is carried inside.  At the same time, there will be a .json suffix file.  It is used to import via software and modify key functions in real time.  https://get.vial.today.  The above URL is vial, which is also a way to change the key. Vial does not need a .json suffix file, and it can be automatically recognized.  I can realize the basic matrix keyboard, but I can't change the key.

CN：这款键盘不能实现实时改键吗？www.caniusevia.com。上面网址是via软件官网。.hex后缀的左右手两个固件用于烧录，里面携带via。同时会有一个.json后缀文件。用于导入via软件，实时修改按键功能。https://get.vial.today。上面网址是vial，这也是一种改键方式，vial不需要.json后缀文件，它可以自动识别。基础矩阵键盘我能实现，但是改键方面我不会。

**Reply by Schievel1** [reply]
Sorry, but this uses plain old QMK. I also think that VIA will not work on this thing, given that the Arduino Micro has only a few hundred bytes left of its memory.

**Reply by Gao0130** [reply]
I took a look at it and it was very complicated. If you use vs code to write the firmware, you need to rewrite it on the basis of your project file, and you need to add a bunch of codes to enable the via function.  Then just write a json.  At present, there is no relevant information about via supporting trackball.  See if you want to study it with me.
